crates/common/src/provider/runtime_transport.rs

```diff
@@ -216,14 +216,11 @@ impl RuntimeTransport {
                 *inner = Some(this.connect().await.map_err(TransportErrorKind::custom)?)
             }
 
-            let mut inner = this.inner.write().await;
             // SAFETY: We just checked that the inner transport exists.
-            let inner_mut = inner.as_mut().expect("We should have an inner transport.");
-
-            match inner_mut {
-                InnerTransport::Http(http) => http.call(req),
-                InnerTransport::Ws(ws) => ws.call(req),
-                InnerTransport::Ipc(ipc) => ipc.call(req),
+            match this.inner.read().await.as_ref().unwrap().clone() {
```

Comment:
we're acquiring this twice now we could do the this combined in an if let else 

by checking `this.inner.read().await.cloned()`

also the write check still has a race condition, so we also need to check for `is_none` there again

Reply:
I don't think we even need to clone since Service is implemented for `&Http` and `&Pubsub`

Reply:
ah right, it works as long as we can obtain &mut &T

```rust
match this.inner.read().await.as_ref().unwrap() {
    InnerTransport::Http(http) => {
        let mut http = http;
        http.call(req)
    },
}
```

is there a cleaner way to do this without that strange let?

Reply:
Don't think so

Reply:
refactored it, now we don't clone and check second time after aquiring write lock

we can't use if let some there anymore I believe because we don't clone

Reply:
can do `InnerTransport::Http(mut http)`?

Reply:
> can do InnerTransport::Http(mut http)?

`mut http` removes the reference

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2278e1efe9c495f4ea5932f229a0a919